### ISSUE TITLE
Warning in bindgen: found non-foreign-function-safe member in struct marked #[repr(C)]

### ISSUE THREAD

**Issue by siscia** [servo/rust-bindgen#634]
### Input C/C++ Header

```C++
#include "src/CDeps/Redis/include/redismodule.h"
#include "src/CDeps/Redis/include/export_redismodule.h"
```

### Bindgen Invokation
```Rust
    #[derive(Debug)]
    struct SqliteTypeChooser;

    impl ParseCallbacks for SqliteTypeChooser {
        fn int_macro(&self, _name: &str, value: i64) -> Option<IntKind> {
            if value >= i32::min_value() as i64 &&
               value <= i32::max_value() as i64 {
                Some(IntKind::I32)
            } else {
                None
            }
        }
    }


    let bindings = bindgen::Builder::default()
        .no_unstable_rust()
        .parse_callbacks(Box::new(SqliteTypeChooser))
        .header("wrapper.h")
        .generate()
        .expect("Unable to generate bindings");

    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings.write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}
```

### Actual Results

and

```rust

extern "C" {
    pub fn fgetpos(__stream: *mut FILE, __pos: *mut fpos_t)
     -> ::std::os::raw::c_int;
}
extern "C" {
    pub fn fsetpos(__stream: *mut FILE, __pos: *const fpos_t)
     -> ::std::os::raw::c_int;
}

```

```
warning: found non-foreign-function-safe member in struct marked #[repr(C)]: found struct without foreign-function-safe representation annotation in foreign module, consider adding a #[repr(C)] attribute to the type, #[warn(improper_ctypes)] on by default
    --> /home/simo/rediSQL-rst/target/debug/build/rediSQL-rst-196441589fefb3b1/out/bindings.rs:5575:48
     |
5575 |     pub fn fgetpos(__stream: *mut FILE, __pos: *mut fpos_t)
     |                                                ^^^^^^^^^^^

warning: found non-foreign-function-safe member in struct marked #[repr(C)]: found struct without foreign-function-safe representation annotation in foreign module, consider adding a #[repr(C)] attribute to the type, #[warn(improper_ctypes)] on by default
    --> /home/simo/rediSQL-rst/target/debug/build/rediSQL-rst-196441589fefb3b1/out/bindings.rs:5579:48
     |
5579 |     pub fn fsetpos(__stream: *mut FILE, __pos: *const fpos_t)
     |                                                ^^^^^^^^^^^^^

```

-->

### Expected Results

I wasn't expecting the warning generated.

### `RUST_LOG=bindgen` Output

It actually didn't change anything, but I it may be me running it wrong.

The project is here: https://github.com/siscia/rediSQL-rs and the C dependencies are here: https://github.com/siscia/rediSQL-rs/tree/master/src/CDeps/Redis

For anything please just let me know :)

**Comment by emilio** [comment]
Could you post the definition of `fpos_t`? May it be an empty struct or something like that?

**Comment by siscia** [comment]
Hi,

Actually all the code I get is the one that you see. I have no idea of the definition of fpos_t however I assume it the one documented here: http://www.cplusplus.com/reference/cstdio/fpos_t/

Stuff from the stdio...

Sorry for this vague information but I really don't know how to help futher...

**Comment by emilio** [comment]
So I can't reproduce it on my system.

The warning points to a source file (in this case `/home/simo/rediSQL-rst/target/debug/build/rediSQL-rst-196441589fefb3b1/out/bindings.rs`). That file should contain the rust definition for `fpos_t`, could you paste that bit, or upload the file?

Thanks again! :)

**Comment by siscia** [comment]
Hi emilio,

what I got is here:

```Rust
#[repr(C)]
#[derive(Debug, Copy)]
pub struct _G_fpos_t {
    pub __pos: __off_t,
    pub __state: __mbstate_t,
}
#[test]
fn bindgen_test_layout__G_fpos_t() {
    assert_eq!(::std::mem::size_of::<_G_fpos_t>() , 16usize , concat ! (
               "Size of: " , stringify ! ( _G_fpos_t ) ));
    assert_eq! (::std::mem::align_of::<_G_fpos_t>() , 8usize , concat ! (
                "Alignment of " , stringify ! ( _G_fpos_t ) ));
    assert_eq! (unsafe {
                & ( * ( 0 as * const _G_fpos_t ) ) . __pos as * const _ as
                usize } , 0usize , concat ! (
                "Alignment of field: " , stringify ! ( _G_fpos_t ) , "::" ,
                stringify ! ( __pos ) ));
    assert_eq! (unsafe {
                & ( * ( 0 as * const _G_fpos_t ) ) . __state as * const _ as
                usize } , 8usize , concat ! (
                "Alignment of field: " , stringify ! ( _G_fpos_t ) , "::" ,
                stringify ! ( __state ) ));
}
impl Clone for _G_fpos_t {
    fn clone(&self) -> Self { *self }
}
#[repr(C)]
#[derive(Debug, Copy)]
pub struct _G_fpos64_t {
    pub __pos: __off64_t,
    pub __state: __mbstate_t,
}

...

pub type fpos_t = _G_fpos_t;
```

I looked inside the source file of Redis and there is nothing like fpos_t however it includes stdio.h

However let me guess that a lot of other projects include stdio.h

Did you try to fork my repo and reproduce the issues or simply recreate the same environment give the indication that I provide to you?

**Comment by emilio** [comment]
> Did you try to fork my repo and reproduce the issues or simply recreate the same environment give the indication that I provide to you?

I forked and built, but couldn't repro. This is inside system headers so that's somewhat expected I guess.

Could you post the definitions of `__off64_t` and `__mbstate_t`?

**Comment by dimbleby** [comment]
duplicate #442?

**Comment by emilio** [comment]
Maybe if the rustc version the reporter is compiling with is doesn't contain https://github.com/rust-lang/rust/pull/39462?

But even with that, I don't see off-hand a reason those types should have `PhantomData`. Oh, perhaps `__mbstate_t` is/contains a union? That'd explain it I guess.

**Comment by siscia** [comment]
Hi,

here are the definitions:

```rust
pub type __off_t = ::std::os::raw::c_long;
pub type __off64_t = ::std::os::raw::c_long;

...

#[repr(C)]
#[derive(Debug, Copy)]
pub struct __mbstate_t {
    pub __count: ::std::os::raw::c_int,
    pub __value: __mbstate_t__bindgen_ty_1,
}
#[repr(C)]
#[derive(Debug, Copy)]
pub struct __mbstate_t__bindgen_ty_1 {
    pub __wch: __BindgenUnionField<::std::os::raw::c_uint>,
    pub __wchb: __BindgenUnionField<[::std::os::raw::c_char; 4usize]>,
    pub bindgen_union_field: u32,
}
```

Is this enough? I see primitive so I would rule out that the structs are empty.

Also, let me post the whole file in this [gist](https://gist.github.com/siscia/a65c505a89b50b0e30a842f530c70c59)

**Comment by siscia** [comment]
Also, I am running/compiling in ubuntu 16.04 with:

```
$ cargo -V
cargo-0.17.0-nightly (f9e5481 2017-03-03)
$ rustc -V
rustc 1.16.0 (30cf806ef 2017-03-10)
```

**Comment by emilio** [comment]
Hmm, interesting. So it seems definitely like a dupe of #442. Given the date you should have the fix in tree though, could you check just in case that updating `rustc` still warns?

**Comment by dimbleby** [comment]
I don't believe that the compiler fix is in stable (1.16.0) rust yet.

**Comment by emilio** [comment]
That's right, closing this as a dupe of #442. Thanks both :)
